BUG: Fix qcut for nullable integers

dsaxton opened the pull request:
- [x] tests added / passed
- [x] passes `black pandas`
- [x] whatsnew entry

Related to #31389 but doesn't close any issues that I'm aware of. Seems the problems were indexing into a numpy array with a BooleanArray and using searchsorted in the presence of pd.NA.

Commit:
Add test

Commit:
Cast BooleanArray to bool

Commit:
Update 1.1 whatsnew

Commit:
Update test

Commit:
Use place holder for NA values

Commit:
Issue number

jschendel commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -200,9 +200,7 @@ Sparse
--
--
-
+- Bug where :meth:`qcut` would raise when passed a nullable integer. (:issue:`31389`)
```
I think this can go in the 1.0.1 whatsnew now that it's been created.

jschendel commented on pandas/core/reshape/tile.py:
```diff
@@ -391,10 +391,13 @@ def _bins_to_cuts(
 
     side = "left" if right else "right"
-    ids = ensure_int64(bins.searchsorted(x, side=side))
+    ids = ensure_int64(bins.searchsorted(np.where(~isna(x), x, -1), side=side))
```
For `cut` we handle this by converting in the `cut` method:
https://github.com/pandas-dev/pandas/blob/882dbee0b781696db06b2dad00bf3b80eda729f7/pandas/core/reshape/tile.py#L210-L214

It looks like we can do this is a similar place within the `qcut` method, and I think it makes more sense to do that so we avoid doing the `np.where` in cases where it's not needed.

WillAyd replied:
Wouldn't this approach also not work for data where -1 is valid?

dsaxton replied:
@jschendel Yes, that makes more sense. @WillAyd I think the NA values get masked later so the place holder you use can be arbitrary.

Commit:
Cast to numpy

Commit:
Merge branch 'master' into qcut-null-int

Commit:
Move note

Commit:
Fixup

Commit:
Don't check for BooleanArray

Commit:
Make test data deterministic

Don't want this to raise with "unlucky" test data that doesn't
have high enough cardinality

jreback commented on pandas/tests/arrays/test_integer.py:
```diff
@@ -1074,6 +1074,17 @@ def test_cut(bins, right, include_lowest):
     tm.assert_categorical_equal(result, expected)
 
 
+@pytest.mark.parametrize("q", [2, 5, 10])
```
can you move to pandas/tests/reshape/test_qcut.py

we group tests by function if possible

jreback commented on pandas/tests/arrays/test_integer.py:
```diff
@@ -1074,6 +1074,17 @@ def test_cut(bins, right, include_lowest):
+    arr = pd.array(np.arange(100), dtype=any_nullable_int_dtype)
+    arr[::2] = pd.NA
+
+    result = pd.qcut(arr, q)
```
do we have an equivalent test for cut?

dsaxton replied:
Yes, there's `test_cut` also in `test_integer.py`. Would you recommend moving / renaming into `tests/reshape/test_cut.py`?

jreback replied:
yes

jreback replied:
> Yes, there's `test_cut` also in `test_integer.py`. Would you recommend moving / renaming into `tests/reshape/test_cut.py`?

there IS already a pandas/tests/reshape/test_cut.py

you can just move the tests

jreback commented on pandas/core/reshape/tile.py:
```diff
@@ -340,6 +340,9 @@ def qcut(
     x = _preprocess_for_cut(x)
     x, dtype = _coerce_to_type(x)
 
+    if is_extension_array_dtype(x.dtype) and is_integer_dtype(x.dtype):
```
this needs to go in _coerce_to_type

dsaxton replied:
So if I put this into _coerce_to_type the qcut tests pass but not the cut tests, because it also changes behavior here: https://github.com/pandas-dev/pandas/blob/master/pandas/core/reshape/tile.py#L208. @TomAugspurger might be able to shed some light on why x needs to be object dtype in cut?

jreback replied:
move this code: https://github.com/pandas-dev/pandas/blob/master/pandas/core/reshape/tile.py#L210

inside _coerce_to_type and might fix all

dsaxton replied:
This does seem to work, but only if we modify the cut test to operate on float arrays instead of object arrays with NaN. Seems something odd might be happening with cut (this is master):

```python
In [5]: a = np.arange(100).astype(object)                                                                                                                                       

In [6]: a[::2] = np.nan                                                                                                                                                         

In [7]: a                                                                                                                                                                       
Out[7]: 
array([nan, 1, nan, 3, nan, 5, nan, 7, nan, 9, nan, 11, nan, 13, nan, 15,
       nan, 17, nan, 19, nan, 21, nan, 23, nan, 25, nan, 27, nan, 29, nan,
       31, nan, 33, nan, 35, nan, 37, nan, 39, nan, 41, nan, 43, nan, 45,
       nan, 47, nan, 49, nan, 51, nan, 53, nan, 55, nan, 57, nan, 59, nan,
       61, nan, 63, nan, 65, nan, 67, nan, 69, nan, 71, nan, 73, nan, 75,
       nan, 77, nan, 79, nan, 81, nan, 83, nan, 85, nan, 87, nan, 89, nan,
       91, nan, 93, nan, 95, nan, 97, nan, 99], dtype=object)

In [8]: object_result = pd.cut(a, 3)                                                                                                                                            

In [9]: float_result = pd.cut(a.astype(float), 3)                                                                                                                               

In [10]: assert_categorical_equal(object_result, float_result)   
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-10-4469237ad337> in <module>
----> 1 assert_categorical_equal(object_result, float_result)

~/pandas/pandas/_testing.py in assert_categorical_equal(left, right, check_dtype, check_category_order, obj)
    843         assert_index_equal(left.categories, right.categories, obj=f"{obj}.categories")
    844         assert_numpy_array_equal(
--> 845             left.codes, right.codes, check_dtype=check_dtype, obj=f"{obj}.codes",
    846         )
    847     else:

~/pandas/pandas/_testing.py in assert_numpy_array_equal(left, right, strict_nan, check_dtype, err_msg, check_same, obj)
   1000     # compare shape and values
   1001     if not array_equivalent(left, right, strict_nan=strict_nan):
-> 1002         _raise(left, right, err_msg)
   1003 
   1004     if check_dtype:

~/pandas/pandas/_testing.py in _raise(left, right, err_msg)
    994             diff = diff * 100.0 / left.size
    995             msg = f"{obj} values are different ({np.round(diff, 5)} %)"
--> 996             raise_assert_detail(obj, msg, left, right)
    997 
    998         raise AssertionError(err_msg)

~/pandas/pandas/_testing.py in raise_assert_detail(obj, message, left, right, diff)
    927         msg += f"\n[diff]: {diff}"
    928 
--> 929     raise AssertionError(msg)
    930 
    931 

AssertionError: Categorical.codes are different

Categorical.codes values are different (33.0 %)
[left]:  [-1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0]
[right]: [-1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 0, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 1, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2, -1, 2]
```

Looks like the result for object dtype is off:

```python
In [16]: a[-5:]                                                                                                                                                                 
Out[16]: array([95, nan, 97, nan, 99], dtype=object)

In [17]: object_result[-5:]                                                                                                                                                     
Out[17]: 
[(0.902, 33.667], NaN, (0.902, 33.667], NaN, (0.902, 33.667]]
Categories (3, interval[float64]): [(0.902, 33.667] < (33.667, 66.333] < (66.333, 99.0]]

In [18]: float_result[-5:]                                                                                                                                                      
Out[18]: 
[(66.333, 99.0], NaN, (66.333, 99.0], NaN, (66.333, 99.0]]
Categories (3, interval[float64]): [(0.902, 33.667] < (33.667, 66.333] < (66.333, 99.0]]
```

Without NaN we get the same output:

```python
In [21]: a = np.arange(100).astype(object)                                                                                                                                      

In [22]: object_result = pd.cut(a, 3)                                                                                                                                           

In [23]: float_result = pd.cut(a.astype(float), 3)                                                                                                                              

In [24]: assert_categorical_equal(object_result, float_result)                                                                                                                  

In [25]:       
```

jreback reviewed (changes requested):
ping on green.

jreback commented:
pls merge master as well

Commit:
Update test scripts

Commit:
Merge branch 'master' into qcut-null-int

Commit:
Move test_cut for nullable int

Commit:
Just import pandas

Commit:
Cast to float instead of object in test

Commit:
Move conversion into _coerce_to_type

Commit:
Merge branch 'master' into qcut-null-int

jreback commented on pandas/core/reshape/tile.py:
```diff
@@ -203,16 +203,13 @@ def cut(
     # NOTE: this binning code is changed a bit from histogram for var(x) == 0
 
     # for handling the cut for datetime and timedelta objects
```
ok, can you move *all* of these comments *inside* _coerce_to_type

jreback reviewed (changes requested):
lgtm. can you move the comments, ping on green.

Commit:
Move comments

jreback commented:
great long on green

Commit:
Remove empty file

dsaxton commented:
@jreback All green, thanks for the review

jreback commented:
thanks @dsaxton
